pkgs/build-support/rust/build-rust-package/default.nix

```diff
@@ -86,22 +75,31 @@ let
 
   sysroot = callPackage ./sysroot { } {
     inherit target shortTarget;
-    RUSTFLAGS = args.RUSTFLAGS or "";
-    originalCargoToml = src + /Cargo.toml; # profile info is later extracted
+    RUSTFLAGS = finalAttrs.RUSTFLAGS or "";
+    originalCargoToml = finalAttrs.src + /Cargo.toml; # profile info is later extracted
   };
 
 in
 
-# Tests don't currently work for `no_std`, and all custom sysroots are currently built without `std`.
-# See https://os.phil-opp.com/testing/ for more information.
-assert useSysroot -> !(args.doCheck or true);
 
-stdenv.mkDerivation ((removeAttrs args [ "depsExtraArgs" "cargoUpdateHook" "cargoLock" ]) // lib.optionalAttrs useSysroot {
-  RUSTFLAGS = "--sysroot ${sysroot} " + (args.RUSTFLAGS or "");
+previousAttrs:
+
+lib.optionalAttrs useSysroot {
+  RUSTFLAGS =
+    # Tests don't currently work for `no_std`, and all custom sysroots are currently built without `std`.
+    # See https://os.phil-opp.com/testing/ for more information.
+    assert useSysroot -> !(finalAttrs.doCheck or true);
+    "--sysroot ${sysroot} " + (finalAttrs.RUSTFLAGS or "");
 } // {
-  inherit buildAndTestSubdir cargoDeps;
+  removeFromBuilderEnv = [ "depsExtraArgs" "cargoUpdateHook" "cargoLock" ];
+
+  inherit cargoDeps;
 
-  cargoBuildType = buildType;
+  buildAndTestSubdir = previousAttrs.buildAndTestSubdir or null;
+
+  cargoBuildType =
+    assert buildType == "release" || buildType == "debug";
```

Comment:
Can we move the assserts after the inputs? thats a bit cleaner than inlining them.

Reply:
Due to the evaluation model of `finalAttrs`, the assertions can not be kept in the place they were before. Citing from the PR description:

> Asserts involving attributes from `finalAttrs` have to be "hidden" inside of values. Right now, I moved them to the respective attributes/variables; another option would be to put them into a dedicated `asserts` attribute, i.e. `asserts = assert (all asserts here); "";`.